lib/ts/recipe/session/sessionFunctions.ts

```diff
@@ -83,87 +72,73 @@ export async function getSession(
+         * if it comes here, it means token verification has failed.
+         * It may be due to:
+         *  - signing key was updated and this token was signed with new key
+         *  - access token is actually expired
+         *  - access token was signed with the older signing key
+         *
+         * if access token is actually expired, we don't need to call core and
+         * just return TRY_REFRESH_TOKEN to the client
+         *
+         * if access token creation time is after this signing key was created
+         * we need to call core as there are chances that the token
+         * was signed with the updated signing key
+         *
+         * if access token creation time is before oldest signing key was created,
+         * so if foundASigningKeyThatIsOlderThanTheAccessToken is still false after
+         * the loop we just return TRY_REFRESH_TOKEN
+         */
+        let payload = parsedAccessToken.payload;
 
-            const timeCreated = sanitizeNumberInput(payload.timeCreated);
-            const expiryTime = sanitizeNumberInput(payload.expiryTime);
+        const timeCreated = sanitizeNumberInput(payload.timeCreated);
+        const expiryTime = sanitizeNumberInput(payload.expiryTime);
 
-            if (expiryTime === undefined || expiryTime < Date.now()) {
-                throw err;
-            }
+        if (expiryTime === undefined || timeCreated == undefined) {
+            throw err;
+        }
 
-            if (timeCreated === undefined) {
+        if (parsedAccessToken.version < 3) {
+            if (expiryTime < Date.now()) {
                 throw err;
             }
-
-            // If we reached a key older than the token and failed to validate the token,
-            // that means it was signed by a key newer than the cached list.
-            // In this case we go to the server.
-            if (timeCreated >= key.createdAt) {
-                foundASigningKeyThatIsOlderThanTheAccessToken = true;
-                break;
+            if (timeCreated <= Date.now() - 60000) {
+                throw err;
             }
+        } else {
+            // Since v3 (and above) tokens contain a kid we can trust the cache-refresh mechanism of the jose library.
+            // This means we do not need to call the core since the signature wouldn't pass verification anyway.
+            throw err;
         }
     }
 
-    // If the token was created before the oldest key in the cache but hasn't expired, then a config value must've changed.
-    // E.g., the access_token_signing_key_update_interval was reduced, or access_token_signing_key_dynamic was turned on.
-    // Either way, the user needs to refresh the access token as validating by the server is likely to do nothing.
-    if (!foundASigningKeyThatIsOlderThanTheAccessToken) {
-        throw new STError({
-            message: "Access token has expired. Please call the refresh API",
-            type: STError.TRY_REFRESH_TOKEN,
-        });
-    }
```

Comment:
won't this check be required for version <= 2 of the access token?

Reply:
This is basically replaced by the max-age check